pkg/feeds/npm/npm.go

```diff
@@ -176,70 +179,35 @@ func fetchAllPackages(registryURL string) ([]*feeds.Package, []error) {
 		uniquePackages[pkg.Title]++
 	}
 
-	// Start a collection of workers to fetch all the packages.
-	// This limits the number of concurrent requests to avoid flooding the NPM
-	// registry API with too many simultaneous requests.
-	workChannel := make(chan struct {
-		pkgTitle string
-		count    int
-	})
-
-	// Define the fetcher function that grabs the repos from NPM
-	fetcherFn := func(pkgTitle string, count int) {
-		pkgs, err := fetchPackage(registryURL, pkgTitle)
-		if err != nil {
-			if !errors.Is(err, errUnpublished) {
-				err = feeds.PackagePollError{Name: pkgTitle, Err: err}
-			}
-			errChannel <- err
-			return
-		}
-		// Apply count slice, guard against a given events corresponding
-		// version entry being unpublished by the time the specific
-		// endpoint has been processed. This seemingly happens silently
-		// without being recorded in the json. An `event` could be logged
-		// here.
-		if len(pkgs) > count {
-			packageChannel <- pkgs[:count]
-		} else {
-			packageChannel <- pkgs
-		}
-	}
+	for pkgTitle, count := range uniquePackages {
+		go func(pkgTitle string, count int) {
+			// Before requesting, wait, so all the requests don't arrive at once.
+			jitter := time.Duration(rand.Intn(maxJitterMillis)) * time.Millisecond //nolint:gosec
+			time.Sleep(jitter)
 
-	// The WaitGroup is used to ensure all the goroutines are complete before
-	// returning.
-	var wg sync.WaitGroup
-
-	// Start the fetcher workers.
-	for i := 0; i < workers; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			for {
-				w, more := <-workChannel
-				if !more {
-					// If we have no more work then return.
-					return
+			pkgs, err := fetchPackage(registryURL, pkgTitle)
+			if err != nil {
+				if !errors.Is(err, errUnpublished) {
+					err = feeds.PackagePollError{Name: pkgTitle, Err: err}
 				}
-				fetcherFn(w.pkgTitle, w.count)
+				errChannel <- err
+				return
 			}
-		}()
+			// Apply count slice, guard against a given events corresponding
+			// version entry being unpublished by the time the specific
+			// endpoint has been processed. This seemingly happens silently
+			// without being recorded in the json. An `event` could be logged
+			// here.
+			if len(pkgs) > count {
+				packageChannel <- pkgs[:count]
+			} else {
+				packageChannel <- pkgs
+			}
+		}(pkgTitle, count)
+		// Pause briefly between each request so all the requests don't arrive at once.
+		time.Sleep(50 * time.Millisecond)
```

Comment:
nit: define the pause duration up in the constants

Reply:
Woops - I meant to delete that. I've added a random delay above inside the goroutine.

Reply:
Haha that also solves it